sensorium/training/trainers.py

```diff
@@ -168,6 +252,40 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
         lr_decay_steps=lr_decay_steps,
     ):
 
+        if include_kldivergence and epoch == dec_starting_epoch:
+            # TODO: include hidden dimension
+            kmeans = KMeans(
+                n_clusters=cluster_number, n_init=kmeans_init, random_state=seed
+            )
+            model.train()
+            feature_list = []
+            features_subset = []
+            random_indices = {}
+            # form initial cluster centres
+
+            for k, readout in model.readout.items():
+                features = readout.features.cpu().detach().squeeze().T.numpy()
+                feature_list.append(np.array(features))
+                random_index = np.random.choice(
```

Comment:
this is something that I mentioned before - if you run this code twice - you might get different `random_index`. So (1), fix a seed here as well, (2) give it an option to read `random_index` dictionary from some pre-saved file to have it same across runs

Reply:
I will do that. I guess I will remove the subsamples at the end because the training runs through with the whole feature set. It was more to speed up the debugging process

Reply:
yes, for that sake of making the first things run - this is totally fine:) maybe just add a todo for yourself in the code that we don't forget it later